### ISSUE TITLE
Maven plugin prints (but does NOT fail due to) a ClassNotFoundException: org.apache.commons.io.file.PathUtils

### ISSUE THREAD

**Issue by vorburger** [vorburger/MariaDB4j#488]
_(This message has been edited from an originally wrong interpretion of cause and effect.)_

The `mariaDB4j-maven-plugin` prints the following: 

```
    [INFO] [INFO] --- mariaDB4j-maven-plugin:2.5.0:stop (post-integration-test) @ start-stop ---
    [INFO] [INFO] Stopping MariaDB4j...
    [INFO] [INFO] Program [/tmp/MariaDB4j/base/bin/mysqld, --no-defaults, --console, --skip-grant-tables, --max_allowed_packet=64M, --basedir=/tmp/MariaDB4j/base, --datadir=/tmp/MariaDB4j/data/42757, --port=42757, --socket=/tmp/MariaDB4j.42757.sock] (in working directory /tmp/MariaDB4j/base) just exited, with value 0
    [INFO] [INFO] Successfully destroyed Program [/tmp/MariaDB4j/base/bin/mysqld, --no-defaults, --console, --skip-grant-tables, --max_allowed_packet=64M, --basedir=/tmp/MariaDB4j/base, --datadir=/tmp/MariaDB4j/data/42757, --port=42757, --socket=/tmp/MariaDB4j.42757.sock] (in working directory /tmp/MariaDB4j/base)
    [INFO] [INFO] Database stopped.
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] BUILD SUCCESS
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] Total time:  5.672 s
    [INFO] [INFO] Finished at: 2021-11-28T22:21:48+01:00
    [INFO] [INFO] ------------------------------------------------------------------------
    [INFO] [INFO] cleanupOnExit() ShutdownHook quietly deleting temporary DB data directory: /tmp/MariaDB4j/data/42757
    [INFO] Exception in thread "Shutdown Hook Deletion Thread for Temporary DB /tmp/MariaDB4j/data/42757" java.lang.NoClassDefFoundError: org/apache/commons/io/file/PathUtils
    [INFO]      at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1341)
    [INFO]      at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:324)
    [INFO]      at org.apache.commons.io.FileUtils.deleteQuietly(FileUtils.java:1236)
    [INFO]      at ch.vorburger.mariadb4j.DB$1.run(DB.java:474)
    [INFO] Caused by: java.lang.ClassNotFoundException: org.apache.commons.io.file.PathUtils
    [INFO]      at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
    [INFO]      at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
    [INFO]      ... 4 more
```

~~This should be made to not appear only on `mvn release:prepare`, but for every `mvn package`, so that whatever PR broke this (possibly #477 or #478 or #482 or #483 today) would fail to pass CI for this to be noticed much earlier.~~

@robin-xyzt-ai given your interest in the Maven Plugin in #476, would you be willing to contrib. a PR for this?

~~This blocks #334.~~

**Comment by vorburger** [comment]
The error above actually does not block or have anything to do with `mvn release:prepare` (only #TBD does); it also appears when running `mvn verify` (or just using the plugin in another project, presumably).

This must be some issue with MariaDB4j fixing commons-io to version 2.4.1... but #491 doesn't fix it either.

> This should be made to not appear only on mvn release:prepare, but for every mvn package

That's just because Travis runs `mvn package` which does not include ITs that are only part of `verify`. I've changed that in #490, and Travis currently still passes - and shows `Build Summary: Passed: 3, Failed: 0, Errors: 0, Skipped: 0`.

PS: #489 does not seem to fix this error.

**Comment by vorburger** [comment]
See also #207 for some history about this.

**Comment by robin-xyzt-ai** [comment]
@vorburger I googled a bit and bumped into [this issue](https://issues.redhat.com/browse/JBASMP-61). It sounds like there are workarounds but they are brittle and ugly.

I would propose for this issue to either:
- Replace that call to `FileUtils#deleteQuietly` with calls using only the classes in the standard JDK (basically inlining the code behind that call)
- Or, if you want to keep using `FileUtils#deleteQuietly` as that behavior is tested, wrap that call with a try-catch, and do our own cleanup using standard JDK classes in case of the `ClassNotFoundException`. That way, we only trigger the new code path in tests and not in production.

I can probably provide a PR with the above approach. If you don't like this approach, I'm gonna pass for this issue. Just like you, I don't fancy digging into Maven-related classloading.

**Comment by vorburger** [comment]
> bumped into [this issue](https://issues.redhat.com/browse/JBASMP-61)

Interesting, thanks for having searched and sharing the finding!

> * Replace that call to `FileUtils#deleteQuietly` with calls using only the classes in the standard JDK (basically inlining the code behind that call)

I like that, that's probably the simplest and clearest solution. 

If by inline you essentially mean "copy/paste Apache Commons IO's implementation" then it's not even particularly "risky" anyway.

> * Or, if you want to keep using `FileUtils#deleteQuietly` as that behavior is tested, wrap that call with a try-catch, and do our own cleanup using standard JDK classes in case of the `ClassNotFoundException`. That way, we only trigger the new code path in tests and not in production.
 
That's perhaps overly cautious, and only confusing for maintenance. Also I suspect (but have not personally tested) that this problem actually doesn't just affect the Maven plugin's integration test, but even real world usage of the plugin in full projects? Because if this ONLY shows up during the local build of MariaDB4j itself due to some Classloader issues in Maven Plugin integration tests themselves, but not when using MariaDB4j Maven plugin as an end-user, then perhaps it's less of a priority alltogether.

> I can probably provide a PR with the above approach.

Go for it, I'd be happy to review and merge a PR about this.

**Comment by robin-xyzt-ai** [comment]
To my surprise, creating my own `FileUtils` class which contains all the necessary code just results in

```
[INFO] Exception in thread "Shutdown Hook Deletion Thread for Temporary DB /tmp/MariaDB4j/data/46521" java.lang.NoClassDefFoundError: ch/vorburger/mariadb4j/FileUtils
[INFO]  at ch.vorburger.mariadb4j.DB$1.run(DB.java:474)
[INFO] Caused by: java.lang.ClassNotFoundException: ch.vorburger.mariadb4j.FileUtils
[INFO]  at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
[INFO]  at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
```

There clearly isn't much what you can do in a shut down hook. Perhaps it works if I put all the code in a single file,

**Comment by vorburger** [comment]
Oh if the fact that it's called from a Shutdown Hook is the real issue here
(sorry I hadn't entirely grasped that earlier), then perhaps the best
approach is simply to make our Maven Plugin here use MariaDB4j without a
Shutdown Hook at all (I can't remember if the Builder has something to
prevent it, but you could add it), and instead ensure that some Maven
Lifecycle stop kind of event stops the DB server explicitly "at the right
time" (whatever that means exactly)?

**Comment by robin-xyzt-ai** [comment]
The shutdownhook is installed in a protected method of a non-final class with a protected constructor (`DB#cleanupOnExit`). 
As such, people could have created an instance of this class and rely on this method. The documentation even states that it will install a shutdown hook.

Anyway, by being creative I managed to put all the code in a single class, and that seems to get rid of the stacktrace.
That class:

- Contains the actual shutdown hook code. You'll see some suppliers in the constructor to pass in non-final fields.
- Contains all the file deletion code copied from Apache (not a 100% literal copy as I removed some code paths which we didn't trigger)
- Implements FileVisitor as I needed a FileVisitor, and I could not use an inner/anonymous class without bumping into the same issue.

I'll create the PR immediately. Feel free to reject it if you think that instead we need to get rid of the shutdown hook.
